paper/src/main/java/com/denizenscript/denizen/paper/events/EntityShootsBowPaperScriptEventImpl.java

```diff
@@ -0,0 +1,48 @@
+package com.denizenscript.denizen.paper.events;
+
+import com.denizenscript.denizen.events.entity.EntityShootsBowEvent;
+import com.denizenscript.denizen.objects.ItemTag;
+import com.denizenscript.denizen.utilities.DenizenAPI;
+import com.denizenscript.denizencore.objects.ObjectTag;
+import com.denizenscript.denizencore.objects.core.ElementTag;
+import com.denizenscript.denizencore.utilities.CoreUtilities;
+import org.bukkit.Bukkit;
+import org.bukkit.entity.Player;
+
+public class EntityShootsBowPaperScriptEventImpl extends EntityShootsBowEvent {
+
+    @Override
+    public boolean applyDetermination(ScriptPath path, ObjectTag determinationObj) {
+        if (determinationObj instanceof ElementTag && !isDefaultDetermination(determinationObj)) {
+            String determination = determinationObj.toString();
+            String lower = CoreUtilities.toLowerCase(determination);
+            if (lower.equals("keep_item")) {
+                event.setConsumeArrow(false);
+                return true;
+            }
+        }
+        return super.applyDetermination(path, determinationObj);
+    }
+
+    @Override
+    public ObjectTag getContext(String name) {
+        if (name.equals("item")) {
+            return new ItemTag(event.getArrowItem());
+        }
+        return super.getContext(name);
+    }
+
+    @Override
+    public void fire() {
+        super.fire();
+        if (!event.getConsumeArrow() && entity.isPlayer()) {
+            final Player p = entity.getPlayer();
+            Bukkit.getScheduler().scheduleSyncDelayedTask(DenizenAPI.getCurrentInstance(), new Runnable() {
+                @Override
+                public void run() {
+                    p.updateInventory();
+                }
+            }, 1);

```

Comment:
This should be done by the determation method, not `fire` (ie: this shouldn't run from _some other plugin_ changing the value of `ConsumeArrow`)

Reply:
Should this be done right when the determination is applied or after the call to the superclass determination (with a cancelled check to not updateInv twice)?

Reply:
Right after applied realistically. You can safely assume that there won't be conflicting determinations, in large part because the updateInventory isn't likely to have negative side effects if it ends up being undone.
(After the super call wouldn't really work, as determinations are applied 1-by-1 anyway).